server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java

```diff
@@ -131,6 +132,69 @@ private void update(MessageId messageId, UpdateMessagePatch updateMessagePatch,
 
     }
 
+    private void assertValidUpdate(List<MessageResult> messagesToBeUpdated, UpdateMessagePatch updateMessagePatch, MailboxSession session) throws MailboxException {
+        List<MailboxId> draftMailboxes = mailboxIdFor(Role.DRAFTS, session);
+        List<Flags> futureFlags = patchFlags(messagesToBeUpdated, updateMessagePatch);
+        List<MailboxId> targetMailboxes = getTargetedMailboxes(messagesToBeUpdated, updateMessagePatch);
+
+        boolean targetHasDraft = targetMailboxes.stream().anyMatch(draftMailboxes::contains);
+        boolean targetHasNonDraft = targetMailboxes.stream().anyMatch(id -> !draftMailboxes.contains(id));
+
+        assertValidUpdate(futureFlags, targetHasDraft, targetHasNonDraft);
+    }
+
+    private void assertValidUpdate(List<Flags> futureFlags, boolean targetHasDraft, boolean targetHasNonDraft) throws DraftMessageMailboxUpdateException {
+        assertNoMessageIntersetingDraftMailboxes(targetHasDraft, targetHasNonDraft);
+        assertNoNonDraftMessageInsideDraftMailbox(futureFlags, targetHasDraft);
+        assertNoDraftMessageOutOfDraftMailbox(futureFlags, targetHasNonDraft);
+    }
+
+    private void assertNoMessageIntersetingDraftMailboxes(boolean targetHasDraft, boolean targetHasNonDraft) throws DraftMessageMailboxUpdateException {
+        if (targetHasDraft && targetHasNonDraft) {
+            throw new DraftMessageMailboxUpdateException();
```

Comment:
the exceptions could be more precise, no?

Reply:
Proposition: 

DraftStatusException + explanatory string, the error handling can then re-use the explanation.

Reply:
why not. I thought a class hierarchy would be simpler to handle, no?

Reply:
Hmmm I think it makes it more boiler plate without being more explicit.

If I can't name it well, then it don't deserve to be that special...